src/Traits/HasRoles.php

```diff
@@ -84,6 +84,59 @@ public function scopeRole(Builder $query, $roles): Builder
     }
 
     /**
+     * Scope the model query to certain permissions only.
+     *
+     * @param \Illuminate\Database\Eloquent\Builder $query
+     * @param string|array|\Spatie\Permission\Contracts\Permission|\Illuminate\Support\Collection $permissions
+     *
+     * @return \Illuminate\Database\Eloquent\Builder
+     */
+    public function scopePermission(Builder $query, $permissions): Builder
+    {
+        if ($permissions instanceof Collection) {
+            $permissions = $permissions->toArray();
+        }
+
+        if (! is_array($permissions)) {
+            $permissions = [$permissions];
+        }
+
+        $rolesWithPermissions = collect([]);
+        $permissions = array_map(function ($permission) use (&$rolesWithPermissions) {
```

Comment:
This could be cleaned up greatly. Right now this map operation does two things. Making sure the permissions are models and building up that `rolesWithPermissions`. Do this in two steps (with two map operations I guess) for clarity (so when can use that nasty `&`)